Friend/firmware/firmware_v1.0/src/transport.c

```diff
@@ -143,34 +143,153 @@ void button_pressed(const struct device *dev, struct gpio_callback *cb,
+            //Also do nothing, but transition to the next state
+            current_button_state = ONE_PRESS;
+        }
+
     }
-    else {
-        if(consecutive_hold_count< 20) {
-            bt_gatt_notify(current_connection, &button_service.attrs[1], &consecutive_hold_count, sizeof(consecutive_hold_count));
+
+
+    else if (current_button_state == ONE_PRESS) {
+
+        if (state_ == 0) {
+            inc_count_0++; //button is unpressed
+
+            if (inc_count_0 > 2) {
+                //If button is not pressed for a little while....... 
+                //transition to Two_press. button could be a single or double tap
+                current_button_state = TWO_PRESS;
+                reset_inc_count_0();
+                reset_inc_count_1();            
+            }
         }
-        consecutive_hold_count = 0;
+        if (state_ == 1) {
+
+            inc_count_1++; //button is pressed
+
+            if (inc_count_1 > 50) {
+                //If button is pressed for a long time.......
+                final_button_state = 3;
+                bt_gatt_notify(current_connection, &button_service.attrs[1], &final_button_state, sizeof(final_button_state));
+                printk("long tap\n");
+                //Fire the long mode notify and enter a grace period
+                current_button_state = GRACE;
+                reset_inc_count_0();
+                reset_inc_count_1();
+            }
+
+        }
+
     }
-    if(consecutive_hold_count == 20) {
-    bt_gatt_notify(current_connection, &button_service.attrs[1], &consecutive_hold_count, sizeof(consecutive_hold_count));
+
+    else if (current_button_state == TWO_PRESS) {
+
+        if (state_ == 0) {
+             inc_count_0++; //not pressed
+             if (inc_count_0 > 10) { //10 instances of unpressing
+               
+                if (inc_count_1 > 0) { // if button has been pressed......
+
+                final_button_state = 2;
+                bt_gatt_notify(current_connection, &button_service.attrs[1], &final_button_state, sizeof(final_button_state));
+                printk("double tap\n");
+                //Fire the notify and enter a grace period
+                current_button_state = GRACE;
+                reset_inc_count_0();
+                reset_inc_count_1();
+                
+             }
+             //single button press
+            else {
+                final_button_state = 1;
+                bt_gatt_notify(current_connection, &button_service.attrs[1], &final_button_state, sizeof(final_button_state));
+                printk("single tap\n");
+                //Fire the notify and enter a grace period
+                current_button_state = GRACE;
+                reset_inc_count_0();
+                reset_inc_count_1();
+
+             }
+
+             }
+
+
+        }
+        else if (state_ == 1 ) {
+            inc_count_1++;
+
+            if (inc_count_1 > threshold) {
+                final_button_state = 2;
```

Comment:
Can you change the internal state constants to macros? So instead of 2, it actually reads what the state is. Will make it easier to maintain the logic down the road.